riscv/src/continuations/memory_merkle_tree.rs

```diff
@@ -94,58 +128,154 @@ impl<T: FieldElement, const N_LEAVES_LOG: usize, const WORDS_PER_PAGE: usize>
-            self.hashes[level][index] = hash_cap0(
-                &self.hashes[level + 1][index * 2],
-                &self.hashes[level + 1][index * 2 + 1],
-            );
+            self.update_inner_hash(level, index);
         }
     }
 
+    fn update_leaf_hash(&mut self, page_index: usize) {
+        self.hashes.insert(
+            (N_LEVELS - 1, page_index),
+            Self::hash_page(&self.data[&page_index]),
+        );
+    }
+
+    fn update_inner_hash(&mut self, level: usize, index: usize) {
+        let new_hash = hash_cap0(
+            self.get_hash(level + 1, index * 2),
+            self.get_hash(level + 1, index * 2 + 1),
+        );
+        self.hashes.insert((level, index), new_hash);
+    }
+
     /// Returns the root hash of the Merkle tree.
     pub fn root_hash(&self) -> &[T; 4] {
-        &self.hashes[0][0]
+        self.get_hash(0, 0)
     }
 
     /// Returns the data and Merkle proof for a given page.
     pub fn get(&self, page_index: usize) -> (&[T; WORDS_PER_PAGE], &[T; 4], Vec<&[T; 4]>) {
         let mut proof = vec![];
-        for (level, index) in self.iter_path(page_index).take(N_LEAVES_LOG) {
+        for (level, index) in self.iter_path(page_index).take(N_LEVELS - 1) {
             let sibling_index = index ^ 1;
-            proof.push(&self.hashes[level][sibling_index]);
+            proof.push(self.get_hash(level, sibling_index));
         }
-        assert_eq!(proof.len(), N_LEAVES_LOG);
+        assert_eq!(proof.len(), N_LEVELS - 1);
+
+        let page_hash = self.get_hash(N_LEVELS - 1, page_index);
 
-        let page_hash = &self.hashes[N_LEAVES_LOG][page_index];
+        let page_data = self.data.get(&page_index).unwrap_or(&self.zero_page);
 
-        (&self.data[page_index], page_hash, proof)
+        (page_data, page_hash, proof)
     }
 
     /// Yields (level, index) pairs for the path from the given page to the root.
     fn iter_path(&self, page_index: usize) -> impl Iterator<Item = (usize, usize)> {
-        (0..=N_LEAVES_LOG).rev().map(move |level| {
-            let index = page_index >> (N_LEAVES_LOG - level);
+        (0..N_LEVELS).rev().map(move |level| {
+            let index = page_index >> (N_LEVELS - level - 1);
             (level, index)
         })
     }
+
+    /// For each MerkleTree generic parameters combination, gets the default
+    /// hash value for every level of the tree.
+    ///
+    /// It is completely unfeasible to write a const function to calculate this
+    /// in Rust, also Rust doesn't allow us to have a different static variable
+    /// per monomorphic instantiation of the MerkleTree, so these solutions are
+    /// out.
+    ///
+    /// Instead, the only way I found to cache this calculation is via the
+    /// unsafe erasing of the type of the result array, and then storing it in a
+    /// map indexed by the type of the field element.
+    fn default_hashes_per_level() -> &'static [[T; 4]; N_LEVELS] {
```

Comment:
I think this function can go and instead we can just have `calculate_default_hashes_per_level() -> [[T; 4]; N_LEVELS]`.

Unless I'm understanding something wrong, this function should just compute a small amount of hashes (64 to hash the zero page + 23 to hash all the way to the root). If this happens once for each Merkle tree instantiations, I don't see the need for this complexity.